tokio-util/src/sync/task_tracker.rs

```diff
@@ -0,0 +1,591 @@
+        // `set_closed` since it already uses AcqRel.
+        //
+        // This synchronizes with the release store of any other call to `drop_task`, and with the
+        // release store in the call to `set_closed`. That ensures that everything that happened
+        // before those other calls to `drop_task` or `set_closed` will be visible after this load,
+        // and those things will also be visible to anything woken by the call to `notify_waiters`.
+        self.state.load(Ordering::Acquire);
+
+        self.on_last_exit.notify_waiters();
+    }
+}
+
+impl TaskTracker {
+    /// Create a new `TaskTracker`.
+    ///
+    /// The `TaskTracker` will start out as open.
+    pub fn new() -> Self {
+        Self {
+            inner: Arc::new(TaskTrackerInner::new()),
+        }
+    }
+
+    /// Waits until this `TaskTracker` is both closed and empty.
+    ///
+    /// The `wait` future will resolve if the `TaskTracker` has been both closed and empty at the
+    /// same time at any point since the `wait` future was created. This is the case even if more
+    /// tasks have been spawned by the time that the `wait` future is polled.
+    #[inline]
+    pub fn wait(&self) -> TaskTrackerWaitFuture<'_> {
+        TaskTrackerWaitFuture {
+            future: self.inner.on_last_exit.notified(),
+            inner: if self.inner.is_done() {
+                None
+            } else {
+                Some(&self.inner)
+            },
+        }
+    }
+
+    /// Close this `TaskTracker`.
+    ///
+    /// This allows `wait` futures to complete. It does not prevent you from spawning new tasks.
+    ///
+    /// Returns `true` if this closed the `TaskTracker`, or `false` if it was already closed.
+    #[inline]
+    pub fn close(&self) -> bool {
+        self.inner.set_closed()
+    }
+
+    /// Reopen this `TaskTracker`.
+    ///
+    /// This prevents `wait` futures from completing even if the `TaskTracker` is empty.
+    ///
+    /// Returns `true` if this reopened the `TaskTracker`, or `false` if it was already open.
+    #[inline]
+    pub fn reopen(&self) -> bool {
+        self.inner.set_open()
+    }
+
+    /// Returns true if this `TaskTracker` is closed.
+    #[inline]
+    pub fn is_closed(&self) -> bool {
+        (self.inner.state.load(Ordering::Acquire) & 1) != 0
+    }
+
+    /// Returns true if this `TaskTracker` is open.
+    #[inline]
+    pub fn is_open(&self) -> bool {
+        !self.is_closed()
+    }
+
+    /// Returns the number of tasks tracked by this `TaskTracker`.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.inner.state.load(Ordering::Acquire) >> 1
+    }
+
+    /// Returns true if there are no tasks in this `TaskTracker`.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.inner.state.load(Ordering::Acquire) <= 1
+    }
+
+    /// Spawn the provided future on the Tokio runtime, and track it in this `TaskTracker`.
```

Comment:
it would be nice if this documentation maybe stated that it's just a call to `track_future` wrapped in a spawn, to make it clearer that you can track tasks spawned through other mechanisms (e.g. on a `JoinSet`).